### PR TITLE
Fixed serialization of top level types inheriting from types serializ…

### PR DESCRIPTION
**Pull request by jvbsl**
…able by the serializer

* Rewrote base serializer to actually use the best matching serializer logic, instead of a separated implementation of a property serializer
* Removed VisitInfo, as only the TypeSymbol of it was in use
* Propagate included serializers mask through the serializers to be able to exclude serializers for the top level "this"/"ret"
* In MemberInfo.FullName exclude "this." as a prefix
* In turn refactored the base serialization logic to be less redundant

### CODE REVIEW

File: NonSucking.Framework.Serialization/Serializers/KnownSimpleTypeSerializer.cs [hunk 1]
```diff
@@ -135,16 +137,14 @@ private static KnownTypes GetKnownType(ISymbol typeSymbol)
...
-                    var addressFamilySer =
-                        NoosonGenerator.GenerateStatementsForProps(
-                            new[] { addressFamilyProp }, context, MethodType.Serialize);
-                    foreach (var item in addressFamilySer)
-                    {
-                        statements.MergeWith(item);
-                    }
+                    var addressFamilySer = NoosonGenerator.GenerateStatementsForMember(addressFamilyProp, context,
+                        MethodType.Serialize);
+                    if (addressFamilySer == null)
```
Comment (susch19) [inline NonSucking.Framework.Serialization/Serializers/KnownSimpleTypeSerializer.cs]:
```suggestion
                    if (addressFamilySer is null)
```

File: NonSucking.Framework.Serialization/Serializers/KnownSimpleTypeSerializer.cs [hunk 2]
```diff
@@ -230,9 +230,10 @@ private static KnownTypes GetKnownType(ISymbol typeSymbol)
                             "AddressFamily",
                             propName);
 
-                    var addressFamilySer =
-                        NoosonGenerator.GenerateStatementsForProps(
-                            new[] { addressFamilyProp }, context, MethodType.Deserialize).Single();
+                    var addressFamilySer = NoosonGenerator.GenerateStatementsForMember(addressFamilyProp, context,
+                        MethodType.Deserialize);
+                    if (addressFamilySer == null)
```
Comment (susch19) [inline NonSucking.Framework.Serialization/Serializers/KnownSimpleTypeSerializer.cs]:
```suggestion
                    if (addressFamilySer is null)
```

File: NonSucking.Framework.Serialization/Serializers/PublicPropertySerializer.cs [hunk 3]
```diff
@@ -39,14 +39,13 @@ var props
...
-                   props
-                       .Select(x => x with { Parent = property.FullName })
-                       .ToArray(),
-                   context,
-                   MethodType.Serialize
-               ).SelectMany(x => x.ToMergedBlock()));
+            foreach (var prop in OrderProps(props))
+            {
+                var propCode = GenerateStatementsForMember(prop, context, MethodType.Serialize);
+                if (propCode == null)
```
Comment (susch19) [inline NonSucking.Framework.Serialization/Serializers/PublicPropertySerializer.cs]:
```suggestion
                if (propCode is null)
```

File: NonSucking.Framework.Serialization/Serializers/PublicPropertySerializer.cs [hunk 4]
```diff
@@ -72,7 +71,18 @@ private static IEnumerable<MemberInfo> FilterPropsForNotWriteOnly(IEnumerable<Me
             return props;
         }
 
-        internal static bool TryDeserialize(MemberInfo property, NoosonGeneratorContext context, string readerName, GeneratedSerializerCode statements, int baseTypesLevelProperties = int.MaxValue)
+        private static IEnumerable<MemberInfo> OrderProps(IEnumerable<MemberInfo> props)
+        {
+            return props.OrderBy(x =>
+                                 {
+                                     var attr = x.Symbol.GetAttribute(AttributeTemplates.Order);
+                                     if (attr == null)
```
Comment (susch19) [inline NonSucking.Framework.Serialization/Serializers/PublicPropertySerializer.cs:79]:
```suggestion
                                    
```

File: NonSucking.Framework.Serialization/Serializers/PublicPropertySerializer.cs [hunk 5]
```diff
@@ -72,7 +71,18 @@ private static IEnumerable<MemberInfo> FilterPropsForNotWriteOnly(IEnumerable<Me
...
+                                         return int.MaxValue;
```
Comment (susch19) [inline NonSucking.Framework.Serialization/Serializers/PublicPropertySerializer.cs:80]:
```suggestion
                                         
```

File: NonSucking.Framework.Serialization/Serializers/PublicPropertySerializer.cs [hunk 6]
```diff
@@ -72,7 +71,18 @@ private static IEnumerable<MemberInfo> FilterPropsForNotWriteOnly(IEnumerable<Me
...
+                                     return (int)attr.ConstructorArguments[0].Value!;
```
Comment (susch19) [inline NonSucking.Framework.Serialization/Serializers/PublicPropertySerializer.cs:81]:
```suggestion
                                     return (int)attr?.ConstructorArguments[0].Value ?? int.MaxValue;
```

File: NonSucking.Framework.Serialization/Serializers/PublicPropertySerializer.cs [hunk 7]
```diff
@@ -93,16 +103,14 @@ var props
 
             props = FilterPropsForNotWriteOnly(props);
 
+            foreach (var prop in OrderProps(props))
+            {
+                var propCode = GenerateStatementsForMember(prop, context, MethodType.Deserialize);
+                if (propCode == null)
```
Comment (susch19) [inline NonSucking.Framework.Serialization/Serializers/PublicPropertySerializer.cs]:
```suggestion
                if (propCode is null)
```

**Review by susch19 — DISMISSED** [review]
Looks good, but the retun in Order by can be shortened and than you wouldn't even need the attr variable and it could be made into one line.
Rest ist just the typical == null and is null :)

File: NonSucking.Framework.Serialization/Serializers/PublicPropertySerializer.cs (hunk 6 above)
Comment (susch19) [reply NonSucking.Framework.Serialization/Serializers/PublicPropertySerializer.cs:81]: Best case new: 17,48357685 ticks
Worst case old: 0,00000925 ticks

**Review by susch19 — DISMISSED** [review]

File: NonSucking.Framework.Serialization/Serializers/KnownSimpleTypeSerializer.cs [hunk 8]
```diff
@@ -115,12 +116,13 @@ private static KnownTypes GetKnownType(ISymbol typeSymbol)
     }
 
     internal static bool TrySerialize(MemberInfo property, NoosonGeneratorContext context, string writerName,
-        GeneratedSerializerCode statements)
+        GeneratedSerializerCode statements, NoosonGenerator.SerializerMask includedSerializers)
```
Comment (Gallimathias) [inline NonSucking.Framework.Serialization/Serializers/KnownSimpleTypeSerializer.cs]: Warum ist die SerializerMask eigentlich Nested?

File: NonSucking.Framework.Serialization/Serializers/PublicPropertySerializer.cs [hunk 9]
```diff
@@ -93,16 +103,14 @@ var props
 
             props = FilterPropsForNotWriteOnly(props);
 
+            foreach (var prop in OrderProps(props))
+            {
+                var propCode = GenerateStatementsForMember(prop, context, MethodType.Deserialize);
+                if (propCode is null)
+                    continue;
+                statements.Statements.AddRange(propCode.ToMergedBlock());
```
Comment (Gallimathias) [inline NonSucking.Framework.Serialization/Serializers/PublicPropertySerializer.cs]:
```suggestion
                var propCode = GenerateStatementsForMember(prop, context, MethodType.Deserialize);
                
                if (propCode is null)
                    continue;
                    
                statements.Statements.AddRange(propCode.ToMergedBlock());
```

**Review by Gallimathias — CHANGES_REQUESTED** [review]

**Review by Gallimathias — APPROVED** [review]

**Pull request merged**